### PR TITLE
Fix username collision

### PR DESCRIPTION
**Pull request by Ph0tonic**
This is a PR to document the solution proposed here https://github.com/jupyterhub/oauthenticator/pull/459 and to discuss it.

### CODE REVIEW

**Comment by sgaist** [comment]
@Ph0tonic Thanks for the PR !

I have done some improvements with regards to the prefix handling.

@manics can you take a look at the latest implementation ?

**Comment by manics** [comment]
There's an extreme corner case where if the login service name has a colon in this would behave incorrectly. Perhaps the easiest solution is to require that `username_prefix` can't contain a `:`, and raise an exception if it does?

**Comment by sgaist** [comment]
@manics do you mean not use any separator between the prefix and the original username ?

**Comment by manics** [comment]
Sorry, I meant ensure that **`login_service`** doesn't contain a  `:`- this should ensure the first `:` is the separator between the authenticator and the username

**Comment by sgaist** [comment]
@manics no worries, it's all good now.

File: multiauthenticator/multiauthenticator.py [hunk 1]
```diff
@@ -69,12 +79,46 @@ def __init__(self, *arg, **kwargs):
...
+                    )
+
             service_name = authenticator_configuration.pop("service_name", None)
 
             authenticator = WrapperAuthenticator(**authenticator_configuration)
 
-            if service_name:
+            if service_name is not None:
+                if service_name.endswith(PREFIX_SEPARATOR):
+                    raise ValueError(f"Service name cannot end with {PREFIX_SEPARATOR}")
```
Comment (manics) [inline multiauthenticator/multiauthenticator.py:118]:
```suggestion
                if PREFIX_SEPARATOR in service_name:
                    raise ValueError(f"Service name cannot contain {PREFIX_SEPARATOR}")
```

File: multiauthenticator/multiauthenticator.py [hunk 2]
```diff
@@ -69,12 +79,46 @@ def __init__(self, *arg, **kwargs):
...
                 authenticator.service_name = service_name
+            elif authenticator.login_service.endswith(PREFIX_SEPARATOR):
+                raise ValueError(f"Login service cannot end with {PREFIX_SEPARATOR}")
```
Comment (manics) [inline multiauthenticator/multiauthenticator.py:121]:
```suggestion
            elif PREFIX_SEPARATOR in authenticator.login_service:
                raise ValueError(f"Login service cannot contain {PREFIX_SEPARATOR}")
```

File: multiauthenticator/tests/test_multiauthenticator.py [hunk 3]
```diff
@@ -184,3 +188,107 @@ def test_extra_configuration():
...
+            },
+        ),
+    ]
+
+    GitLabOAuthenticator.login_service = "test me:"
+
+    with pytest.raises(ValueError) as excinfo:
+        MultiAuthenticator()
+
+    assert f"Login service cannot end with {PREFIX_SEPARATOR}" in str(excinfo.value)
```
Comment (manics) [inline multiauthenticator/tests/test_multiauthenticator.py:294]:
```suggestion
    assert f"Login service cannot contain {PREFIX_SEPARATOR}" in str(excinfo.value)
```

Comment (sgaist) [reply multiauthenticator/tests/test_multiauthenticator.py]: Fixed

File: multiauthenticator/multiauthenticator.py (hunk 2 above)
Comment (sgaist) [reply multiauthenticator/multiauthenticator.py]: Fixed

File: multiauthenticator/multiauthenticator.py (hunk 1 above)
Comment (sgaist) [reply multiauthenticator/multiauthenticator.py]: Fixed

**Comment by sgaist** [comment]
@manics thanks for the suggestions.

I just realized something, I may have over-engineered `service_name`. `login_service` is part of the `OAuthenticator` class and not something than can be modified through traitlets hence the warning that appears during the tests. But it can be set afterward. Thinking about this, I was wondering whether we could not simply re-use that name as well for the other authenticators in the context of the MultiAuthenticator and thus remove the special handling being currently done.

That would be something to be done in a different MR though.

**Comment by manics** [comment]
I can think of arguments both ways. Using an existing property means less configuration, whereas using a new property means admins can customise the resulting username (e.g. they may prefer a short prefix instead of the longer name), and makes it easier to add Authenticators that aren't related to OAuthenticator.

**Comment by sgaist** [comment]
@manics good points, I have refactored and improved the tests with that in mind.

**Pull request merged**